### PR TITLE
Fix setExpiryPolicy for ICache 

### LINKED ISSUE #25547: ICache method setExpiryPolicy with CreatedExpiryPolicy set max duration independently on passed value [HZ-3264]
**Describe the bug**
ICache method `setExpiryPolicy` suppose to set a new expiry policy for the given key in the cache. However it seems it's currently ignored (tried with `CreatedExpiryPolicy`) and once it's called the given entry is never removed from the cache (instead of using the new policy).

**Expected behavior**
New policy should be applied for the entry with given key in the cache.

**To Reproduce**

Use `cache.setExpiryPolicy` as following:
```
ICache<String, String> cache = client.getCacheManager().getCache("hz/myCache");
cache.put("k", "v", new CreatedExpiryPolicy(new Duration(TimeUnit.MILLISECONDS, 5000)));
cache.setExpiryPolicy("k", new CreatedExpiryPolicy(new Duration(TimeUnit.MILLISECONDS, 6000)));
```
and check that entry "k" is never removed from the cache. If the `cache.setExpiryPolicy()` is not called then entry expires in 5 seconds.

### PR DESCRIPTION
**Pull request by gareth-johnston**
Issue #25547 found that it is possible to update the expiry policy for a ICache entry and it would never get removed in-line with the provided policy. The issue appears to be that the update to the records policy accessed the expiry-for-update time (which is null) rather than creation time, which meant the record is updated with the value `-1` for expiry time. 

Added unit tests to cover the reported failure case. 

Fixes #25547 

Checklist:
- [X] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [X] Add `Add to Release Notes` label if changes should be mentioned in release notes or `Not Release Notes content` if changes are not relevant for release notes
- [X] Request reviewers if possible

### CODE REVIEW

**Review by JackPGreen — COMMENTED** [review]
I've read the Javadoc of `getExpiryForUpdate`/`getExpiryForCreation` a couple of times, but I'm not super clear on the distinction.

But shouldn't `com.hazelcast.cache.impl.AbstractCacheRecordStore.updateRecordWithExpiry(Data, Object, R, ExpiryPolicy, long, boolean, int, UUID, UUID)` be changed to be consistent?

**Comment by gareth-johnston** [comment]
> getExpiryForUpdate

Good spot, thanks. 

I don't believe we use the 'ExpiryForUpdate' time - just the 'ExpiryForCreation' to determine when to expire for cache. Perhaps @vbekiaris can weigh in.

**Review by JackPGreen — APPROVED** [review]

File: hazelcast/src/test/java/com/hazelcast/cache/CacheBasicAbstractTest.java [hunk 1]
```diff
@@ -1044,6 +1045,26 @@ public void cacheManagerOfCache_cannotBeOverwrittenConcurrently() throws Excepti
...
     }
 
+    @Test
+    public void testUpdatingExpiryPolicyOfCacheEntry() {
+        ICache<Integer, String> cache = createCache();
+        cache.put(1, "value", new CreatedExpiryPolicy(new Duration(TimeUnit.MILLISECONDS, 5000)));
+
+        assertTrue(cache.setExpiryPolicy(1, new CreatedExpiryPolicy(new Duration(TimeUnit.MILLISECONDS, 6000))));
+        assertTrueEventually("expected the key to be expired", () ->
+                Assertions.assertNull(cache.get(1)), 10);
```
Comment (JamesHazelcast) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheBasicAbstractTest.java:1055]: Nit-pick but you're using the jupiter extension `Assertions` in a legacy JUnit test class - I'd prefer to see `Assert` used here

**Review by JamesHazelcast — COMMENTED** [review]
Thanks for picking this up @gareth-johnston.

Besides a couple of small comments below, my main concern relies around being thorough with our solution. Ondrej describes the initial issue by saying that `setExpiryPolicy` is being ignored entirely, adding that he only tested with `CreatedExpiryPolicy`, and I'd like to make sure we're confident the solution does not introduce unexpected behaviour for other policies. Did you run tests against any other policies? Do you think it's worth including more expiry policies for coverage?

I haven't looked into the implementation details of `getExpiryForUpdate` vs `getExpiryForCreation` etc, but seeing those methods being interchanged as a solution, I want to make sure we're not missing anything. 👍

File: hazelcast/src/test/java/com/hazelcast/cache/CacheBasicAbstractTest.java (hunk 1 above)
Comment (JackPGreen) [reply hazelcast/src/test/java/com/hazelcast/cache/CacheBasicAbstractTest.java:1055]: > Nit-pick but you're using the jupiter extension `Assertions` in a legacy JUnit test class - I'd prefer to see `Assert` used here

Good spot - I don't think you can mix&match like that.

**Comment by gareth-johnston** [comment]
@JackPGreen @JamesHazelcast FYI closing this as on closer inspection I have determined that this functionality is in-line with the API contract. Will post more details on the original ticket.

**Pull request closed without merging**